### ISSUE TITLE
compile_error in const fn with condition

### ISSUE THREAD

**Issue by burrbull** [rust-lang/rust#79888]
This code can't be compiled.
```rust
pub const fn new() {
    if false {
        compile_error!("unreachable compile time error");
    }
}
```
https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=0cccbd21f7671757acd8112152ce916d

I think that such behaviour is wrong.

**Comment by jonas-schievink** [comment]
This works as expected. `compile_error!` causes an error during macro expansion.
